pkg/ddl/backfilling.go

```diff
@@ -787,11 +793,58 @@ func (dc *ddlCtx) runAddIndexInLocalIngestMode(
 	return bcCtx.FinishAndUnregisterEngines(ingest.OptCleanData | ingest.OptCheckDup)
 }
 
-func executeAndClosePipeline(ctx *OperatorCtx, pipe *operator.AsyncPipeline) error {
+func adjustWorkerPoolSize(ctx context.Context, pipe *operator.AsyncPipeline, job *model.Job, avgRowSize int) {
+	opR, opW := pipe.GetLocalIngestModeReaderAndWriter()
+	if opR == nil || opW == nil {
+		logutil.DDLIngestLogger().Error("failed to get local ingest mode reader or writer", zap.Int64("jobID", job.ID))
+		return
+	}
+	reader, readerOk := opR.(*TableScanOperator)
+	writer, writerOk := opW.(*IndexIngestOperator)
+	if !readerOk || !writerOk {
+		logutil.DDLIngestLogger().Error(
+			"unexpected operator types, config can't be adjusted",
+			zap.Int64("jobID", job.ID),
+			zap.Bool("isReaderValid", readerOk),
+			zap.Bool("isWriterValid", writerOk),
+		)
+		return
+	}
+	ticker := time.NewTicker(UpdateDDLJobReorgCfgInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			targetReaderCnt, targetWriterCnt := expectedIngestWorkerCnt(job.ReorgMeta.GetConcurrency(), avgRowSize)
+			currentReaderCnt, currentWriterCnt := reader.GetWorkerPoolSize(), writer.GetWorkerPoolSize()
+			if int32(targetReaderCnt) == currentReaderCnt && int32(targetWriterCnt) == currentWriterCnt {
+				continue
+			}
+			reader.TuneWorkerPoolSize(int32(targetReaderCnt))
+			writer.TuneWorkerPoolSize(int32(targetWriterCnt))
+			logutil.DDLIngestLogger().Info("adjust ddl job config success",
+				zap.Int64("jobID", job.ID),
+				zap.Int32("table scan operator count", reader.GetWorkerPoolSize()),
+				zap.Int32("index ingest operator count", writer.GetWorkerPoolSize()))
+		}
+	}
+}
+
+func executeAndClosePipeline(ctx *OperatorCtx, pipe *operator.AsyncPipeline, job *model.Job, avgRowSize int) error {
 	err := pipe.Execute()
 	if err != nil {
 		return err
 	}
+
+	// Adjust worker pool size dynamically.
+	if job != nil && avgRowSize != 0 {
+		go func() {
+			adjustWorkerPoolSize(ctx, pipe, job, avgRowSize)
+		}()
+	}
```

Comment:
```suggestion
	// Adjust worker pool size dynamically.
	if job != nil {
		go func() {
			adjustWorkerPoolSize(ctx, pipe, job, avgRowSize)
		}()
	}
```

`expectedIngestWorkerCnt` can handle `avgRowSize == 0` correctly. BTW, is it possible that `job` is nil?

Reply:
we only pass the non nil `job` from `runAddIndexInLocalIngestMode()`, and if the job is nil, the error will be reported in the previous of `runAddIndexInLocalIngestMode()`.

Reply:
Here we judge whether the job is nil to decide if it's a local ingest job, if not, we don't start the goroutine.